community/values/src/main/java/org/neo4j/values/storable/DurationValue.java

```diff
@@ -518,55 +518,41 @@ static DurationValue durationBetween( Temporal from, Temporal to )
             }
         }
         // Compute the time difference - which is simple at this point
-        return difference( months, days, from, to );
+        // NANOS of a day will never overflow a long
+        long nanos = from.until( to, NANOS );
+        return newDuration( months, days, nanos / NANOS_PER_SECOND, (int) (nanos % NANOS_PER_SECOND) );
     }
 
-    private static DurationValue difference( long months, long days, Temporal from, Temporal to )
+    private static DurationValue durationInSecondsAndNanos( Temporal from, Temporal to )
     {
-        long seconds = 0;
-        long nanos = 0;
-        if ( from.isSupported( SECONDS ) )
+        long seconds;
+        long nanos;
+        boolean negate = false;
+        if ( from.isSupported( OFFSET_SECONDS ) && !to.isSupported( OFFSET_SECONDS ) )
         {
-            if ( to.isSupported( SECONDS ) )
-            {
-                boolean negate = false;
-                if ( from.isSupported( OFFSET_SECONDS ) && !to.isSupported( OFFSET_SECONDS ) )
-                {
-                    negate = true;
-                    Temporal tmp = from;
-                    from = to;
-                    to = tmp;
-                }
-                seconds = from.until( to, SECONDS );
-                nanos = to.getLong( NANO_OF_SECOND ) - from.getLong( NANO_OF_SECOND );
-                if ( seconds > 0 && nanos < 0 )
-                {
-                    seconds--;
-                    nanos = NANOS_PER_SECOND - nanos;
-                }
-                else if ( seconds < 0 && nanos > 0 )
-                {
-                    seconds++;
-                    nanos = nanos - NANOS_PER_SECOND;
-                }
-                if ( negate )
-                {
-                    seconds = -seconds;
-                    nanos = -nanos;
-                }
-            }
-            else
-            {
-                seconds = -from.getLong( SECOND_OF_DAY );
-                nanos = -from.getLong( NANO_OF_SECOND );
-            }
+            negate = true;
+            Temporal tmp = from;
+            from = to;
+            to = tmp;
+        }
+        seconds = from.until( to, SECONDS );
+        int fromNanos = from.isSupported( NANO_OF_SECOND ) ? from.get( NANO_OF_SECOND ) : 0;
+        int toNanos = to.isSupported( NANO_OF_SECOND ) ? to.get( NANO_OF_SECOND ) : 0;
+        nanos = toNanos - fromNanos;
+        if ( seconds > 0 && nanos < 0 )
+        {
+            nanos = NANOS_PER_SECOND + nanos;
```

Comment:
Why do I get the itching feeling that we should to `seconds--;` here?

Reply:
Convince yourself: Duration from `1.000000005 s` to `3.000000003 s`.
Line 538: `seconds = 2`
Line 541: `nanos = -2`
Line 544: `nanos = 1 000 000 000 - (-2) = 999 999 998
In total: `1.999999998 s`

Same for the `seconds++` below.

Reply:
I got the same feeling and wrote some tests for it. Actually this is handled by the `until` function but I found another bug that occurs sometimes when `seconds` are zero so that I've fixes